test/functional/wallet_keypool_topup.py

```diff
@@ -24,71 +15,58 @@ def add_options(self, parser):
         self.skip_if_no_wallet()
 
     def run_test(self):
-        wallet_path = self.nodes[1].wallets_path / self.default_wallet_name / self.wallet_data_filename
-        wallet_backup_path = self.nodes[1].datadir_path / "wallet.bak"
-        self.generate(self.nodes[0], COINBASE_MATURITY + 1)
-
-        self.log.info("Make backup of wallet")
-        self.stop_node(1)
-        shutil.copyfile(wallet_path, wallet_backup_path)
-        self.start_node(1, self.extra_args[1])
-        self.connect_nodes(0, 1)
-        self.connect_nodes(0, 2)
-        self.connect_nodes(0, 3)
-
-        for i, output_type in enumerate(["legacy", "p2sh-segwit", "bech32"]):
-
+        node = self.nodes[0]
+
+        output_types = ["legacy", "p2sh-segwit", "bech32"]
+        default_wallet = node.get_wallet_rpc(self.default_wallet_name)
+        for output_type in output_types:
+            wallet_name = str(output_type)
+            node.createwallet(wallet_name)
+            wallet = node.get_wallet_rpc(wallet_name)
             self.log.info("Generate keys for wallet with address type: {}".format(output_type))
-            idx = i+1
             for _ in range(90):
-                addr_oldpool = self.nodes[idx].getnewaddress(address_type=output_type)
+                addr_oldpool = wallet.getnewaddress(address_type=output_type)
             for _ in range(20):
-                addr_extpool = self.nodes[idx].getnewaddress(address_type=output_type)
+                addr_extpool = wallet.getnewaddress(address_type=output_type)
 
             # Make sure we're creating the outputs we expect
-            address_details = self.nodes[idx].validateaddress(addr_extpool)
-            if i == 0:
+            address_details = wallet.validateaddress(addr_extpool)
+            if output_type == "legacy":
                 assert not address_details["isscript"] and not address_details["iswitness"]
-            elif i == 1:
+            elif output_type == "p2sh-segwit":
                 assert address_details["isscript"] and not address_details["iswitness"]
             else:
                 assert not address_details["isscript"] and address_details["iswitness"]
 
-
             self.log.info("Send funds to wallet")
-            self.nodes[0].sendtoaddress(addr_oldpool, 10)
-            self.generate(self.nodes[0], 1)
-            self.nodes[0].sendtoaddress(addr_extpool, 5)
-            self.generate(self.nodes[0], 1)
+            default_wallet.send(outputs=[{addr_oldpool:10}, {addr_extpool:5}])
+            self.generate(node, 1)
 
-            self.log.info("Restart node with wallet backup")
-            self.stop_node(idx)
-            shutil.copyfile(wallet_backup_path, wallet_path)
-            self.start_node(idx, self.extra_args[idx])
-            self.connect_nodes(0, idx)
-            self.sync_all()
+            node.unloadwallet(wallet_name)
+            node.loadwallet(wallet_name)
+            wallet = node.get_wallet_rpc(wallet_name)
```

Comment:
In bb54a7ad0cf22686e3f6735213bac1dd4453cb15 "test: use multiple wallets instead of nodes in `wallet_keypool_topup`"

This is incorrect and removing the thing that the test is actually testing for.

The purpose of the test is to ensure that receiving to something in the keypool will cause everything up to that address being removed from the keypool, that the keypool is subsequently topped up, and that an address from past the end of the initial keypool will also be detected after the first topup, and trigger it's own topup. That's why this has a whole thing with creating and restoring a backup - so that there is a wallet with an untouched keypool to do the test on.

Simply unloading and reloading the wallet does not test that the keypool gets topped up correctly when it discovers stuff in the keypool.

Reply:
It needs some more love but I just re-wrote the first commit correcting achow's points: https://github.com/furszy/bitcoin-core/commit/738860039d1f2e2038e64b7d8d0cd23f3e6b88a5.

Got a similar speedup for the descriptors wallet:

```
Before:
./wallet_keypool_topup.py 5.63s user 0.57s system 56% cpu 10.945 total

After:
./wallet_keypool_topup.py 1.69s user 0.19s system 72% cpu 2.605 total
```

But got a slowdown for the legacy wallet:
```
Before:
./wallet_keypool_topup.py --legacy-wallet  9.52s user 2.07s system 32% cpu 35.514 total

After:
./wallet_keypool_topup.py --legacy-wallet  5.03s user 2.15s system 16% cpu 44.458 total
```

@brunoerg, if you like it, cherry-pick it and see if you can clean it further. I did not spend much time on it.

Reply:
Thanks, @furszy. I'm checking it now.

Reply:
@furszy Do you want to address it on a follow-up?

Reply:
> @furszy Do you want to address it on a follow-up?

Sure.